### PR TITLE
 okcoinusd: handleErrors & InsufficientFunds and InvalidOrder exceptions

### PR DESCRIPTION
**Pull request by mkutny**
Tested on okex.

If you uncomment error testing from #1141 this PR will now pass.

Along the way, I explicitly set price min limits - it's much easier to handle them on the client side if they are set.

I also enabled okex keys for testing as they passed all the tests on my side.

### CODE REVIEW

**Comment by mkutny** [comment]
'**' could not be transpiled? I'm positive you added something like that a while ago.

**Comment by kroitor** [comment]
@mkutny, naah, that was about converting the `typeof xxx ===` clause ) It won't be able to convert the switch-case in handleErrors, therefore we do that in `if { } else if {} else {}`. But don't worry, I would clean it anyway upon merging.

**Comment by mkutny** [comment]
Hmm... `switch` would look much more elegant in case of structured error codes.

**Comment by kroitor** [comment]
@mkutny actually, you're right. You know what, I think I'm going to add a few transpilation regexes for switch-cases, if you are ok with waiting for a while (will need a little more time for it, mostly because of other tasks).

**Comment by kroitor** [comment]
@mkutny nope, looks like bad luck: https://stackoverflow.com/questions/60208/replacements-for-switch-statement-in-python, will have to rewrite it for if-else or for a string-map...

**Comment by mkutny** [comment]
I'll redo with if-else, no worries.

**Comment by mkutny** [comment]
I left a FIXME there where you allow exchange-specific and unified param values.

As a rule of thumb, I find it intuitive if we provide unified values to standard parameters and exchange-specific values to `params` object. Those unified values in `params` object got me confused, I had to spend some time reading exchange's API docs.

**Comment by kroitor** [comment]
> As a rule of thumb, I find it intuitive if we provide unified values to standard parameters and exchange-specific values to params object. Those unified values in params object got me confused, I had to spend some time reading exchange's API docs.

I agree, will keep exchange specific ints there only.

**Comment by mkutny** [comment]
I'm confused, transpiler seems to start failing where it was working before.

File: js/okcoinusd.js [hunk 1]
```diff
@@ -153,6 +153,8 @@ module.exports = class okcoinusd extends Exchange {
                 'price': markets[i]['maxPriceDigit'],
             };
             let lot = Math.pow (10, -precision['amount']);
+            let minAmount = markets[i]['minTradeSize'];
+            let minPrice = 1 / (10 ** precision['price']);
```
Comment (kroitor) [inline js/okcoinusd.js]: It cannot transpile the `**`, unfortunately, this is js-only-syntax for now... So, it has to be Math.pow.

File: js/okcoinusd.js [hunk 2]
```diff
@@ -623,17 +623,29 @@ module.exports = class okcoinusd extends Exchange {
...
-        let response = await this.fetch2 (path, api, method, params, headers, body);
+    handleErrors (code, reason, url, method, headers, body) {
+        if (this.verbose)
+            console.log (this.id, method, url, code, reason, body ? ("\nResponse:\n" + body) : '');
+        let response = JSON.parse (body);
+        if ('error_code' in response) {
+            if (this.errorCodes === undefined) {
+                this.errorCodes = {
+                    '1009': OrderNotFound,
+                    '1003': InvalidOrder,      // no order type (was left by previous author)
```
Comment (kroitor) [inline js/okcoinusd.js]: Python PEP8 standard demands that here should be exactly 2 spaces before the in-line comment starts. In JS we only leave one space, and the second space is added by the transpiler to the Python-version of the source.
So, these lines should look like 
```
                         ↓
    '1003': InvalidOrder, // no order type (was left by previous author)
```

Transpiler is in fact very happy, it's the python code syntax check that fails.

Also, I think it may be a good idea to move these errorCodes to `describe()` as that method looks like appropriate place for initializing this property, what do you think?

File: js/okcoinusd.js [hunk 3]
```diff
@@ -623,17 +623,29 @@ module.exports = class okcoinusd extends Exchange {
...
+                    '1002': InsufficientFunds, // The transaction amount exceed the balance
+                    '10000': ExchangeError,    // createLimitBuyOrder(symbol, undefined, undefined)
+                    '10008': ExchangeError,    // Illegal URL parameter
+                };
+            }
+            let Exception = this.errorCodes[response['error_code']];
+            if (Exception)
+                throw new Exception (this.id + ' ' + this.json (response));
+            else
+                throw new ExchangeError (this.id + ' ' + this.json (response));
```
Comment (kroitor) [inline js/okcoinusd.js]: Also, I'm not sure if this particular construct will work well in all languages. Most likely will have to rewrite this for if-else nevertheless. Currently the transpiler can only guess proper throw syntax within one line, no code lookups are done, so it's dump thin converter (for now).

**Comment by kroitor** [comment]
> I'm confused, transpiler seems to start failing where it was working before.

I've put the explanations there, let me know if you will edit those places, otherwise I'll do that today, hopefully, and will merge then. Thx!

File: js/okcoinusd.js (hunk 2 above)
Comment (mkutny) [reply js/okcoinusd.js]: > Also, I think it may be a good idea to move these errorCodes to describe() as that method looks like appropriate place for initializing this property, what do you think?

Actually, I did it in the `describe()` from start, as it was *the right* thing to do. But then, from the practical perspective, I moved it closer to the usage site as it was to unhandy to jump between very end and the beginning of the file.

Comment (kroitor) [reply js/okcoinusd.js]: We are free to move the method itself higher up in the file )

File: js/okcoinusd.js (hunk 3 above)
Comment (mkutny) [reply js/okcoinusd.js]: Oh, my God, it's going to be UGLY. There are just six checks at the moment but as the library is gaining usage we'll be discovering more of them. I won't be surprised if we end up with dozens of cases and those `if`s will span pages of code.

Is it possible to do a helper for all three languages? Kind of "ThrowableMap".

@kornrunner, I see you do a lot of contribution on PHP side. I'm not an expert on that language (actually, I'm not an expert in JS too :), but how would you address this issue?

Basically, there are exchanges that report errors by codes and those would be ideal to handle via a `switch` statement. Considering, we can't use `switch`, how would you address such a task in PHP?

https://github.com/mkutny/ccxt/blob/a43b3fdc22696fb40120174e59e9019098460a4b/js/okcoinusd.js#L630-L645

Comment (kroitor) [reply js/okcoinusd.js]: I can tell right away, this is ok in PHP (variables = exception classes). I'm not sure on Python, in that particular case it will definitely throw, because Exception is a reserved keyword for the base exception class there.

File: js/okcoinusd.js [hunk 4]
```diff
@@ -623,17 +623,29 @@ module.exports = class okcoinusd extends Exchange {
         return { 'url': url, 'method': method, 'body': body, 'headers': headers };
     }
 
-    async request (path, api = 'public', method = 'GET', params = {}, headers = undefined, body = undefined) {
-        let response = await this.fetch2 (path, api, method, params, headers, body);
+    handleErrors (code, reason, url, method, headers, body) {
+        if (this.verbose)
+            console.log (this.id, method, url, code, reason, body ? ("\nResponse:\n" + body) : '');
```
Comment (kroitor) [inline js/okcoinusd.js]: This is the last line it complains about, namely the conditional. Apart from it, looks clean. I'll fix it, you don't have to bother.

Comment (mkutny) [reply js/okcoinusd.js]: Too late )

Actually, I was going to move it to the base class anyway. We can do it once there before calling `haldleErrors` instead of doing it in every impl. Ok?

Comment (kroitor) [reply js/okcoinusd.js]: I support the proposal, less duplication == better.

**Comment by mkutny** [comment]
OMG, all checks have passed. Just can't believe my luck.

**Comment by kroitor** [comment]
@mkutny congrats, you did a fantastic job! )
